sources/api/pluto/src/main.rs

```diff
@@ -270,11 +271,80 @@ fn pause_container_uri(region: &str) -> Result<String> {
     ))
 }
 
+/// Returns the number of CPUs on current instance.
+fn get_num_cpus() -> usize {
+    num_cpus::get()
+}
+
+/// Calculates the amount of memory to reserve for kubeReserved in mebibytes.
+/// This function is taken from EKS' calculations.
+/// https://github.com/awslabs/amazon-eks-ami/blob/db28da15d2b696bc08ac3aacc9675694f4a69933/files/bootstrap.sh#L228-L239
+fn get_memory_mebibytes_to_reserve(client: &Client, session_token: &str) -> Result<String> {
+    let max_num_pods = get_max_pods(&client, &session_token)?;
+    let mebibytes_unit = "Mi";
+    let max_num_pods_u32 = max_num_pods.parse::<u32>().context(error::ParseToU32 {
+        object: max_num_pods,
+    })?;
+    let memory_to_reserve = max_num_pods_u32 * 11 + 255;
+
+    Ok(memory_to_reserve.to_string() + mebibytes_unit)
+}
+
+/// Calculates the amount of CPU to reserve for kubeReserved in millicores
+/// from the total number of vCPUs available on the instance.
+/// We are using these CPU ranges from GKE
+/// (https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-architecture#node_allocatable):
+///  6% of the first core
+///  1% of the next core (up to 2 cores)
+///  0.5% of the next 2 cores (up to 4 cores)
+///  0.25% of any cores above 4 cores
+/// This function is taken from EKS' calculations.
+/// https://github.com/awslabs/amazon-eks-ami/blob/db28da15d2b696bc08ac3aacc9675694f4a69933/files/bootstrap.sh#L241-L251
+fn get_cpu_millicores_to_reserve(num_cpus: usize) -> Result<String> {
+    let total_cpu_on_instance = num_cpus * 1000;
+    let cpu_ranges = vec![0, 1000, 2000, 4000, total_cpu_on_instance];
+    let cpu_percentage_reserved_for_ranges = vec![600, 100, 50, 25];
+    let mut cpu_to_reserve: f32 = 0.0;
+    let millicores_unit = "m";
+
+    for i in 0..(cpu_percentage_reserved_for_ranges.len()) {
+        let start_range = cpu_ranges[i];
+        let end_range = cpu_ranges[i + 1];
+        let percentage_to_reserve_for_range = cpu_percentage_reserved_for_ranges[i];
+
+        cpu_to_reserve += get_resource_to_reserve_in_range(
+            total_cpu_on_instance,
+            start_range,
+            end_range,
+            percentage_to_reserve_for_range,
+        )
+    }
+    Ok(cpu_to_reserve.to_string() + millicores_unit)
+}
```

Comment:
It would be more idiomatic to write this with iterators.

For CPUs, have an iterator that repeats 1000 and take num_cpus elements from it.

For percentages, chain an iterator that produces 0.06, 0.01, 0.005, with an iterator that repeats 0.0025.

Then zip the two together, multiply each pair, and sum up the result.

Reply:
That's a good idea. The reason why I do this way is that I want to make it consistent with `get_cpu_millicores_to_reserve` and `get_resource_to_reserve_in_range` functions in `amazon-eks-ami`, where I got the idea from. Do you think I should optimize it or keep this?

Reply:
Definitely optimize it - the comments from the EKS AMI implementation are valuable, but translating the implementation directly from shell script to Rust makes for awkward code.

Reply:
Yeah see my previous comment here: https://github.com/bottlerocket-os/bottlerocket/pull/1388#discussion_r594746238

The algorithm is very simple and can be done with a match statement and a few constants. I had to request tests first to make sure I even understood it!

I agree with Ben on this, simplify it down to a match statement that returns the correct value.